### PR TITLE
PART 2 - ARC-2491 send config data

### PR DESCRIPTION
**Pull request by joshkay10**
**What's in this PR?**
Added the logic to listen to the plugins save event
extract the data from the incoming xml
send extracted information through new API

Tests

**What's Next?**
Consumption!

### CODE REVIEW

File: src/main/java/com/atlassian/jira/cloud/jenkins/common/model/PluginSiteData.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package com.atlassian.jira.cloud.jenkins.common.model;
+
+import java.util.Optional;
+
```
Comment (joshkay10) [inline src/main/java/com/atlassian/jira/cloud/jenkins/common/model/PluginSiteData.java:4]: Since there can be many sites, created this object so i can loop-de-loop through it when extracting from XML

File: src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java [hunk 2]
```diff
@@ -1,41 +1,123 @@
...
+    @Inject
+    public void setXmlUtils(final XmlUtils xmlUtils) {
+        this.xmlUtils = xmlUtils;
+    }
+
     @Override
     public void onChange(final Saveable saveable, final XmlFile file) {
-        if (!(saveable instanceof com.atlassian.jira.cloud.jenkins.config.JiraCloudPluginConfig)) {
-            return;
+        if (saveable instanceof com.atlassian.jira.cloud.jenkins.config.JiraCloudPluginConfig) {
```
Comment (joshkay10) [inline src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java:50]: onChange event fires a lot, this condition filters it to run on saving the main config page and the xml  arg that is passed is our config section..... whether the plugin details are dirty or not makes no difference.

File: src/main/java/com/atlassian/jira/cloud/jenkins/common/model/PluginSiteData.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+public class PluginSiteData {
```
Comment (rachellerathbone) [inline src/main/java/com/atlassian/jira/cloud/jenkins/common/model/PluginSiteData.java:5]: Can you add some Javadoc comments to this to be consistent with other classes that live in the model dir?

File: src/main/java/com/atlassian/jira/cloud/jenkins/util/XmlUtils.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import java.io.IOException;
+import java.io.StringReader;
+
+public class XmlUtils {
```
Comment (rachellerathbone) [inline src/main/java/com/atlassian/jira/cloud/jenkins/util/XmlUtils.java:16]: Same here: some Javadoc comments will help future devs easily identify the purpose of this class.

File: src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java [hunk 5]
```diff
@@ -1,41 +1,123 @@
...
+                                sendDataToPluginApi(
+                                        data,
+                                        autoBuildsRegex,
+                                        autoBuildsEnabled,
+                                        autoDeploymentsEnabled,
+                                        autoDeploymentsRegex,
+                                        secret));
+            }
+        } catch (Exception e) {
+            logger.log(Level.SEVERE, "Error processing plugin configuration data", e);
```
Comment (rachellerathbone) [inline src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java:81]: I don't think we should throw a runtime exception here as we'll lose info about the original exception and can lead to unexpected behaviour (including potential crashes 😬 ) I just did some reading on common XML errors Maybe we can do this a little more gracefully by doing something like:

```
catch (ParserConfigurationException e) {
        // Handle XML parsing configuration error
        logger.log(Level.SEVERE, "XML parsing configuration error", e);
        // we could just return here or return a default value?
        return;
        } catch (SAXException e) {
        // Handle XML parsing SAX (simple api for xml) error
        logger.log(Level.SEVERE, "XML parsing SAX error", e);
        // same as above
        return;
        } catch (Exception e) {
        // Handle other unexpected exceptions
        logger.log(Level.SEVERE, "Error processing plugin configuration data", e);
        // Custom error handling class (you can call it whatever you want) that would extend from Exception
        throw new XmlProcessingException("Failed to process plugin configuration data", e);
        }
    }
```

File: src/test/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListenerTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                any(PipelineLogger.class),
+                anyString(),
+                eq(true),
+                eq("regex"),
+                eq(true),
+                eq("regex")
+        );
+    }
+
+}
```
Comment (rachellerathbone) [inline src/test/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListenerTest.java:107]: Should we have some more test cases here? Perhaps for when things aren't successful/edge cases?

File: src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java [hunk 7]
```diff
@@ -1,41 +1,123 @@
...
 
-    private static final Logger LOGGER =
-            Logger.getLogger(ConfigurationSaveableListener.class.getName());
+    private PluginConfigApi pluginConfigApi;
+    private SecretRetriever secretRetriever;
+    private XmlUtils xmlUtils;
 
-    private transient PluginConfigApi pluginConfigApi;
+    @Inject
+    public void setSecretRetriever(final SecretRetriever secretRetriever) {
```
Comment (rachellerathbone) [inline src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java:34]: We probably want to make sure the injected dependencies are not null to avoid running into NullPointerExceptions. So I say we chuck something like

```
if (secretRetriever == null) {
        throw new IllegalArgumentException("SecretRetriever cannot be null.");
    }
    ```
    
    in for these 3 injections.

File: src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java (hunk 5 above)
Comment (joshkay10) [reply src/main/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListener.java:99]: Agree with not throwing runtime error for exactly what you said!!!
But since we aren't doing anything meaningful with various exceptions I've kept it generic, all we want to do is log, I've avoided re throwing the error as well for the same reason, if an error occurs we don't want to propagate it up, logging is as much dealing with it as we need and higher functions will just muddy the stacktrace.

File: src/test/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListenerTest.java (hunk 6 above)
Comment (joshkay10) [reply src/test/java/com/atlassian/jira/cloud/jenkins/listeners/ConfigurationSaveableListenerTest.java:205]: added a handful more tests and made a mockdatafunction :excite:

**Review by rachellerathbone — APPROVED** [review]
